# Loading the ROS node model

First we need to load the ROS node model in Imandra. We do that using `System.mod_use` and load the messaging module and the model module: 

In [2]:
System.mod_use ~quiet:true "src-messages/basic_types.ml";;
System.mod_use ~quiet:true "src-messages/std_msgs.ml";;
System.mod_use ~quiet:true "src-messages/rosgraph_msgs.ml";;
System.mod_use ~quiet:true "src-messages/geometry_msgs.ml";;
System.mod_use ~quiet:true "src-messages/sensor_msgs.ml";;

open Ros_model;;
open Ros_message;;

- : unit = ()
- : unit = ()


The model has a `state` datatype and a `one_step : state -> state` transition function.

In [3]:
#show state;;

type nonrec state = {
  mode : Ros_model.mode;
  incoming : Ros_model.incoming option;
  outgoing : Ros_model.outgoing option;
}


The `state` contains two "slots" for incoming and outgoing messages and a `mode` value that describes the current  behaviour mode of our robot:

In [4]:
#show mode;;

type nonrec mode = Stopped | Turning | Driving | Error


Our robot starts in the `Stopped`  mode and then transitions between `Turning` and `Driving` modes, reacting to the sensor inputs. 

In [5]:
one_step;;

- : Ros_model.state -> Ros_model.state = <fun>


# Verifying error mode transition

First thing that we want to verify it that our node will not go into an `Error` mode unless it receives an invalid message.  We can formalize this statement as the following logical expression:

$$ \forall s. ( Mode(s) \ne Error) \,\land\, Valid(IncomingMessage(s)) \,\Rightarrow\, (Mode(one\_step(s)) \ne Error) $$

Meaning that for every state $s$, if the state is not already in the `Error` mode and its incoming message is valid, then the state's mode after we've called `one_step` on it is never the `Error` mode.  

We can almost literally encode this formal excpresson as an Imandra `theorem`:

In [6]:
theorem error_state_invalid_message state =
  match state.incoming with None -> true | Some msg -> 
  (  state.mode <> Error 
  && valid_incoming msg  
  ) ==> (one_step state).mode <> Error


val error_state_invalid_message : Ros_model.state -> bool = <fun>


Proved proof call graph proof summary full graph ground_instances 0 definitions 0 inductions 0 search_time 0.009s details Expand smt_stats (:added-eqs 164
 :conflicts 14
 :datatype-accessor-ax 37
 :datatype-constructor-ax 16
 :datatype-splits 8
 :decisions 21
 :del-clause 19
 :max-memory 16.21
 :memory 10.43
 :minimized-lits 2
 :mk-bool-var 132
 :mk-clause 43
 :num-allocs 527483565
 :num-checks 2
 :propagations 59
 :rlimit-count 67348) Expand start[0.009s]
 if :var_0:.Ros_model.incoming = None then true
 else
 not (:var_0:.Ros_model.mode = Ros_model.Error)
 && (if Is_a(Ros_model.Sensor, Option.get :var_0:.Ros_model.incoming)
 then List.length ….Ros_message.Sensor_msgs.ranges = 5 else true)
 ==> not
 ((if :var_0:.Ros_model.mode = Ros_model.Error
 then
 {Ros_model.mode = :var_0:.Ros_model.mode;
 Ros_model.incoming = :var_0:.Ros_model.incoming;
 Ros_model.outgoing = …}
 else
 if :var_0:.Ros_model.incoming = None
 then
 {Ros_model.mode = :var_0:.Ros_model.mode;
 Ros_model.incoming = :var_0:.Ros_model.incoming;
 Ros_model.outgoing = …}
 else …).Ros_model.mode
 = Ros_model.Error) simplify into (:var_0:.Ros_model.incoming = None
 || not
 (not (:var_0:.Ros_model.mode = Ros_model.Error)
 && (not Is_a(Ros_model.Sensor, Option.get :var_0:.Ros_model.incoming)
 || List.length
 (Destruct(Ros_model.Sensor, 0,
 Option.get :var_0:.Ros_model.incoming)).Ros_message.Sensor_msgs.ranges
 = 5)))
|| not
 ((if :var_0:.Ros_model.mode = Ros_model.Error
 then
 {Ros_model.mode = :var_0:.Ros_model.mode;
 Ros_model.incoming = :var_0:.Ros_model.incoming;
 Ros_model.outgoing = None}
 else
 if :var_0:.Ros_model.incoming = None
 then
 {Ros_model.mode = :var_0:.Ros_model.mode;
 Ros_model.incoming = :var_0:.Ros_model.incoming;
 Ros_model.outgoing = None}
 else
 if not Is_a(Ros_model.Sensor, Option.get :var_0:.Ros_model.incoming)
 || List.length
 (Destruct(Ros_model.Sensor, 0,
 Option.get :var_0:.Ros_model.incoming)).Ros_message.Sensor_msgs.ranges
 = 5
 then … else …).Ros_model.mode
 = Ros_model.Error) expansions [] rewrite_steps forward_chaining unsat (let ((a!1 (asserted (not (= (Ros_model.mode_2128 state_2207)
 Ros_model… digraph "proof" {
p_199 [label="Start (if :var_0:.Ros_model.incoming = None then true\l else\l not (:var_0:.Ros_model.mode = Ros_model.Error)\l && (if Is_a(Ros_model.Sensor, Option.get :var_0:.Ros_model.incoming)\l then List.length ….Ros_message.Sensor_msgs.ranges = 5\l else true)\l ==\> not\l ((if :var_0:.Ros_model.mode = Ros_model.Error\l then\l \{Ros_model.mode = :var_0:.Ros_model.mode;\l Ros_model.incoming = :var_0:.Ros_model.incoming;\l Ros_model.outgoing = …\}\l else\l if :var_0:.Ros_model.incoming = None\l then\l \{Ros_model.mode = :var_0:.Ros_model.mode;\l Ros_model.incoming = :var_0:.Ros_model.incoming;\l Ros_model.outgoing = …\}\l else …).Ros_model.mode\l = Ros_model.Error)\l :time 0.009s)",shape=box,style=filled,fontname="courier",fontsize=14];
p_199 -> p_198 [label=""];
p_198 [label="Simplify ((:var_0:.Ros_model.incoming = None\l \|\| not\l (not (:var_0:.Ros_model.mode = Ros_model.Error)\l && (not\l Is_a(Ros_model.Sensor,\l Option.get :var_0:.Ros_model.incoming)\l \|\| List.length\l (Destruct(Ros_model.Sensor, 0,\l Option.get :var_0:.Ros_model.incoming)).Ros_message.Sensor_msgs.ranges\l = 5)))\l \|\| not\l ((if :var_0:.Ros_model.mode = Ros_model.Error\l then\l \{Ros_model.mode = :var_0:.Ros_model.mode;\l Ros_model.incoming = :var_0:.Ros_model.incoming;\l Ros_model.outgoing = None\}\l else\l if :var_0:.Ros_model.incoming = None\l then\l \{Ros_model.mode = :var_0:.Ros_model.mode;\l Ros_model.incoming = :var_0:.Ros_model.incoming;\l Ros_model.outgoing = None\}\l else\l if not\l Is_a(Ros_model.Sensor,\l Option.get :var_0:.Ros_model.incoming)\l \|\| List.length\l (Destruct(Ros_model.Sensor, 0,\l Option.get :var_0:.Ros_model.incoming)).Ros_message.Sensor_msgs.ranges\l = 5\l then … else …).Ros_model.mode\l = Ros_model.Error)\l :expansions [] :rw [] :fc [])",shape=box,style=filled,fontname="courier",fontsize=14];
p_198 -> p_197 [label=""];

One can see that the theorem is "Proven", meaning that Imandra has formally checked that this property holds for all possible input states. 

Let's also verify that in the `Error` mode our model ingnores all the incoming messages:

In [15]:
theorem error_state_ignores_messages state =
  match state.incoming with None -> true | Some msg -> 
  (  state.mode = Error 
  && state.outgoing = None
  ) ==> state.outgoing = None

val error_state_ignores_messages : Ros_model.state -> bool = <fun>


Proved proof call graph proof summary full graph ground_instances 0 definitions 0 inductions 0 search_time 0.010s details Expand smt_stats (:max-memory 17.99
 :memory 17.81
 :mk-bool-var 5
 :num-allocs 617524247
 :rlimit-count 84380) Expand start[0.010s]
 if :var_0:.Ros_model.incoming = None then true
 else
 :var_0:.Ros_model.mode = Ros_model.Error
 && :var_0:.Ros_model.outgoing = None
 ==> :var_0:.Ros_model.outgoing = None simplify into (:var_0:.Ros_model.incoming = None
 || not
 (:var_0:.Ros_model.mode = Ros_model.Error
 && :var_0:.Ros_model.outgoing = None))
|| :var_0:.Ros_model.outgoing = None expansions [] rewrite_steps forward_chaining unsat (let ((a!1 (asserted (not (= (Ros_model.outgoing_2130 state_2327) None_3)))))
 (unit-resolution (as… digraph "proof" {
p_225 [label="Start (if :var_0:.Ros_model.incoming = None then true\l else\l :var_0:.Ros_model.mode = Ros_model.Error\l && :var_0:.Ros_model.outgoing = None\l ==\> :var_0:.Ros_model.outgoing = None\l :time 0.010s)",shape=box,style=filled,fontname="courier",fontsize=14];
p_225 -> p_224 [label=""];
p_224 [label="Simplify ((:var_0:.Ros_model.incoming = None\l \|\| not\l (:var_0:.Ros_model.mode = Ros_model.Error\l && :var_0:.Ros_model.outgoing = None))\l \|\| :var_0:.Ros_model.outgoing = None :expansions [] :rw [] \l :fc [])",shape=box,style=filled,fontname="courier",fontsize=14];
p_224 -> p_223 [label=""];
p_223 -> p_z3_4041 [label="",fontname="courier",fontsize=14];
p_z3_4041 [label="(let ((a!1 (asserted (not (= (Ros_model.outgoing_2130 state_2327) None_3)))))
 (unit-resolution (as…",shape=box,fontname="courier",fontsize=14];p_223 [label="unsat",shape=box,style=filled,fontname="courier",fontsize=14];
}
 Load graph Loading.. digraph "call graph" {
goal [label="not\l((state.Ros_model.incoming = None\l \|\| not\l (state.Ros_model.mode = Ros_model.Error\l && state.Ros_model.outgoing = None))\l \|\| state.Ros_model.outgoing = None)",shape=box,style=filled,color="cyan",fontname="courier",fontsize=14];
}
 Load graph Loading..

# Verifying stopping if facing a wall

Let us verify some more complex statement about our model -- lets prove that if the robot senses that it is too close to the wall, then it always attemps to either stop or backpedal from it. More precisely, if the front-facing sensor shows a range that is too close (say, less than 0.5 m), then our node always sends out a `Twist` message with `linear.x` velocity that is less or equal to zero.

We create two helper functions that get the range of the front sensor and check that the outgoing message is a `Twist` message and its velocity is not positive:

In [7]:
let get_front_sensor_range state =
  match state.incoming with 
  | Some (Sensor sensor) -> List.nth 2 sensor.Sensor_msgs.ranges 
  | _ -> None 

let no_positive_outgoing_velocity state =
  match state.outgoing with 
  | Some (Twist twist) -> twist.Geometry_msg.linear.x <= 0
  | _ -> false

val get_front_sensor_range : Ros_model.state -> Ros_message.number option =
  <fun>
val no_positive_outgoing_velocity : Ros_model.state -> bool = <fun>


Now we can attempt to prove our goal. We require that the state incoming message is a valid `Sensor` message with the front sensor range < 0.5 m, and we want to prove that after we apply `one_call` to the state -- there is always a `Twist` message with non-positive velocity:

In [8]:
verify (fun  state ->
  match state.incoming with None -> true | Some msg -> 
  match get_front_sensor_range state with None -> true | Some range_front ->
  let state' = one_step state in
  (  range_front < 500
  && valid_incoming msg  
  ) ==> 
  no_positive_outgoing_velocity state' 
)

- : Ros_model.state -> bool = <fun>
module CX : sig val state : Ros_model.state end


Refuted call graph proof-attempt digraph "call graph" {
goal [label="not\l(((state.Ros_model.incoming = None\l \|\| (if Is_a(Ros_model.Sensor, Option.get state.Ros_model.incoming)\l && Is_a(Some, state.Ros_model.incoming)\l then\l List.nth 2\l (Destruct(Ros_model.Sensor, 0, Option.get state.Ros_model.incoming)).Ros_message.Sensor_msgs.ranges\l else None)\l = None)\l \|\| not\l (not\l (500 \<=\l Option.get\l (if Is_a(Ros_model.Sensor, Option.get state.Ros_model.incoming)\l && Is_a(Some, state.Ros_model.incoming)\l then\l List.nth 2\l (Destruct(Ros_model.Sensor, 0, Option.get state.Ros_model.incoming)).Ros_message.Sensor_msgs.ranges\l else None))\l && (not Is_a(Ros_model.Sensor, Option.get state.Ros_model.incoming)\l \|\| List.length\l (Destruct(Ros_model.Sensor, 0,\l Option.get state.Ros_model.incoming)).Ros_message.Sensor_msgs.ranges\l = 5)))\l \|\| (Is_a(Ros_model.Twist,\l Option.get\l (if state.Ros_model.mode = Ros_model.Error then … else …).Ros_model.outgoing)\l && Is_a(Some,\l (if state.Ros_model.mode = Ros_model.Error then … else …).Ros_model.outgoing))\l && (Destruct(Ros_model.Twist, 0,\l Option.get\l (if state.Ros_model.mode = Ros_model.Error then … else …).Ros_model.outgoing)).Ros_message.Geometry_msg.linear.Ros_message.x\l \<= 0)",shape=box,style=filled,color="cyan",fontname="courier",fontsize=14];
call_3172 [label="List.nth (-1 + 3)\l(List.tl\l (Destruct(Ros_model.Sensor, 0, Option.get …)).Ros_message.Sensor_msgs.ranges)",shape=box,style=filled,color="green",fontname="courier",fontsize=14];
call_3172 -> expr_3506 [label="="];
expr_3506 [label="if List.tl\l (Destruct(Ros_model.Sensor, 0, Option.get ….Ros_model.incoming)).Ros_message.Sensor_msgs.ranges\l = []\lthen None\lelse\lif -1 + 3 = 0\lthen\l Some\l (List.hd\l (List.tl\l (Destruct(Ros_model.Sensor, 0, Option.get ….Ros_model.incoming)).Ros_message.Sensor_msgs.ranges))\lelse\l List.nth (-1 + -1 + 3)\l (List.tl\l (List.tl\l (Destruct(Ros_model.Sensor, 0, Option.get ….Ros_model.incoming)).Ros_message.Sensor_msgs.ranges))",shape=box,style=filled,color="grey",fontname="courier",fontsize=14];
expr_3175 -> call_3172 [label="calls"];
call_3907 [label="List.nth (-1 + -1 + -1 + 1)\l(List.tl\l (List.tl\l (List.tl\l (Destruct(Ros_model.Sensor, 0, Option.get …)).Ros_message.Sensor_msgs.ranges)))",shape=box,style=filled,color="green",fontname="courier",fontsize=14];
call_3907 -> expr_4436 [label="="];
expr_4436 [label="if List.tl\l (List.tl\l (List.tl\l (Destruct(Ros_model.Sensor, 0, Option.get ….Ros_model.incoming)).Ros_message.Sensor_msgs.ranges))\l = []\lthen None\lelse\lif -1 + -1 + -1 + 1 = 0\lthen\l Some\l (List.hd\l (List.tl\l (List.tl\l (List.tl\l (Destruct(Ros_model.Sensor, 0, Option.get ….Ros_model.incoming)).Ros_message.Sensor_msgs.ranges))))\lelse\l List.nth (-1 + -1 + -1 + -1 + 1)\l (List.tl\l (List.tl\l (List.tl\l (List.tl\l (Destruct(Ros_model.Sensor, 0, Option.get ….Ros_model.incoming)).Ros_message.Sensor_msgs.ranges))))",shape=box,style=filled,color="grey",fontname="courier",fontsize=14];
expr_3909 -> call_3907 [label="calls"];
call_4114 [label="List.nth (-1 + -1 + -1 + -1 + 3)\l(List.tl\l (List.tl\l (List.tl\l (List.tl\l (Destruct(Ros_model.Sensor, 0, Option.get …)).Ros_message.Sensor_msgs.ranges))))",shape=box,style=filled,color="yellow",fontname="courier",fontsize=14];
expr_4116 -> call_4114 [label="calls"];
call_3504 [label="List.nth (-1 + -1 + 3)\l(List.tl\l (List.tl\l (Destruct(Ros_model.Sensor, 0, Option.get …)).Ros_message.Sensor_msgs.ranges))",shape=box,style=filled,color="green",fontname="courier",fontsize=14];
call_3504 -> expr_3839 [label="="];
expr_3839 [label="if List.tl\l (List.tl\l (Destruct(Ros_model.Sensor, 0, Option.get ….Ros_model.incoming)).Ros_message.Sensor_msgs.ranges)\l = []\lthen None\lelse\lif -1 + -1 + 3 = 0\lthen\l Some\l (List.hd\l (List.tl\l (List.tl\l (Destruct(Ros_model.Sensor, 0, Option.get ….Ros_model.incoming)).Ros_message.Sensor_msgs.ranges)))\lelse\l List.nth (-1 + -1 + -1 + 3)\l (List.tl\l (List.tl\l (List.tl\l (Destruct(R

We've failed to prove the theorem, and Imandra gave us a conunter example `CX` module. This module contains a state, that refutes the statement of the theorem we've attempted to prove. Lets check what is the incoming sensor range. And which outgoing message we get if we pass the state into the `one_step`:

In [12]:
get_front_sensor_range CX.state;;
(one_step CX.state).outgoing;;

- : Ros_message.number option = Some 461
- : Ros_model.outgoing option = None


The incoming sensor range is `461` and it is, indeed, less than `500`. But we have no outgoing message -- because the node is in the `Error` mode: 

In [13]:
CX.state.mode

- : Ros_model.mode = Error


In the `Error` mode we are ignoring all the inputs -- so we are not getting any outgoing message (as we've proven in the theorem `error_state_ignores_messages` in the previous section). Adding the requirement that the input state is not in the `Error` mode, we successfully prove the theorem:

In [16]:
theorem stop_if_front_sensor_close state =
  match state.incoming with None -> true | Some msg -> 
  match get_front_sensor_range state with None -> true | Some range_front ->
  let state' = one_step state in
  (  range_front < 500 
  && state.mode <> Error
  && valid_incoming msg  
  ) ==> 
  no_positive_outgoing_velocity state' 

val stop_if_front_sensor_close : Ros_model.state -> bool = <fun>


Proved proof call graph proof summary full graph ground_instances 14 definitions 0 inductions 0 search_time 0.023s details Expand smt_stats (:added-eqs 1634
 :arith-add-rows 13
 :arith-assert-diseq 29
 :arith-assert-lower 82
 :arith-assert-upper 32
 :arith-bound-prop 1
 :arith-conflicts 1
 :arith-eq-adapter 32
 :arith-fixed-eqs 9
 :arith-pivots 7
 :conflicts 60
 :datatype-accessor-ax 91
 :datatype-constructor-ax 141
 :datatype-occurs-check 4805
 :datatype-splits 254
 :decisions 739
 :del-clause 203
 :final-checks 42
 :max-memory 18.77
 :memory 18.77
 :minimized-lits 3
 :mk-bool-var 672
 :mk-clause 221
 :num-allocs 659924009
 :num-checks 30
 :propagations 763
 :rlimit-count 97944) Expand start[0.023s]
 if :var_0:.Ros_model.incoming = None then true
 else
 if (if Is_a(Ros_model.Sensor, Option.get :var_0:.Ros_model.incoming)
 && Is_a(Some, :var_0:.Ros_model.incoming)
 then List.nth 2 ….Ros_message.Sensor_msgs.ranges else None)
 = None
 then true
 else
 Option.get
 (if Is_a(Ros_model.Sensor, Option.get :var_0:.Ros_model.incoming)
 && Is_a(Some, :var_0:.Ros_model.incoming)
 then List.nth 2 ….Ros_message.Sensor_msgs.ranges else None)
 < 500
 && not (:var_0:.Ros_model.mode = Ros_model.Error)
 && (if Is_a(Ros_model.Sensor, Option.get :var_0:.Ros_model.incoming)
 then List.length ….Ros_message.Sensor_msgs.ranges = 5 else true)
 ==> (if Is_a(Ros_model.Twist, Option.get …) && Is_a(Some, …)
 then ….Ros_message.x <= 0 else false) simplify into ((:var_0:.Ros_model.incoming = None
 || (if Is_a(Ros_model.Sensor, Option.get :var_0:.Ros_model.incoming)
 && Is_a(Some, :var_0:.Ros_model.incoming)
 then
 List.nth 2
 (Destruct(Ros_model.Sensor, 0, Option.get :var_0:.Ros_model.incoming)).Ros_message.Sensor_msgs.ranges
 else None)
 = None)
 || not
 ((not
 (500 <=
 Option.get
 (if Is_a(Ros_model.Sensor, Option.get :var_0:.Ros_model.incoming)
 && Is_a(Some, :var_0:.Ros_model.incoming)
 then
 List.nth 2
 (Destruct(Ros_model.Sensor, 0,
 Option.get :var_0:.Ros_model.incoming)).Ros_message.Sensor_msgs.ranges
 else None))
 && not (:var_0:.Ros_model.mode = Ros_model.Error))
 && (not Is_a(Ros_model.Sensor, Option.get :var_0:.Ros_model.incoming)
 || List.length
 (Destruct(Ros_model.Sensor, 0,
 Option.get :var_0:.Ros_model.incoming)).Ros_message.Sensor_msgs.ranges
 = 5)))
|| (Is_a(Ros_model.Twist,
 Option.get
 (if :var_0:.Ros_model.mode = Ros_model.Error then … else …).Ros_model.outgoing)
 && Is_a(Some,
 (if :var_0:.Ros_model.mode = Ros_model.Error then … else …).Ros_model.outgoing))
 && (Destruct(Ros_model.Twist, 0,
 Option.get
 (if :var_0:.Ros_model.mode = Ros_model.Error then …
 else …).Ros_model.outgoing)).Ros_message.Geometry_msg.linear.Ros_message.x
 <= 0 expansions [] rewrite_steps forward_chaining unroll expr (let ((a!1 (Ros_message.Sensor_msgs.ranges_2086
 (get.Ros_model.Sensor.0_2177
 … expansions unroll expr (let ((a!1 (Ros_message.Sensor_msgs.ranges_2086
 (get.Ros_model.Sensor.0_2177
 … expansions unroll expr (let ((a!1 (Ros_message.Sensor_msgs.ranges_2086
 (get.Ros_model.Sensor.0_2177
 … expansions unroll expr (let ((a!1 (Ros_message.Sensor_msgs.ranges_2086
 (get.Ros_model.Sensor.0_2177
 … expansions unroll expr (let ((a!1 (Ros_message.Sensor_msgs.ranges_2086
 (get.Ros_model.Sensor.0_2177
 … expansions unroll expr (let ((a!1 (Ros_message.Sensor_msgs.ranges_2086
 (get.Ros_model.Sensor.0_2177
 … expansions unroll expr (let ((a!1 (Ros_message.Sensor_msgs.ranges_2086
 (get.Ros_model.Sensor.0_2177
 … expansions unroll expr (let ((a!1 (Ros_message.Sensor_msgs.ranges_2086
 (get.Ros_model.Sensor.0_2177
 … expansions unroll expr (let ((a!1 (Ros_message.Sensor_msgs.ranges_2086
 (get.Ros_model.Sensor.0_2177
 … expansions unroll expr (let ((a!1 (Ros_message.Sensor_msgs.ranges_2086
 (get.Ros_model.Sensor.0_2177
 … expansions unroll expr (let ((a!1 (Ros_message.Sensor_msgs.ranges_2086
 (get.Ros_model.Sensor.0_2177
 … expansions unroll expr (let ((a!1 (Ros_message.Sensor_msgs.ranges_2086
 (get.Ros_model.Sensor.0_2177
 … expansions unroll expr (let ((a!1 (Ros_mess

This illustrates how one can use Imandra to iteratively create theorems and prove their correctness or find logical inconsistencies in either the model or our specifications. 